### PR TITLE
Fix edge cases with 1-byte reads

### PR DESCRIPTION
**Pull request by kenkendk**
This PR fixes a missing overload for `ReadByte()` that caused the library to return incorrect results when reading a stream with `ReadByte()`.

As the use of `ReadByte()` on a compressed stream is arguably seldom used, I opted for a simple "read into 1-byte array" solution that is not terribly resource efficient, but prevents having code replicated from the main `Read()` method into `ReadByte()`.

Unfortunately, the main `Read()` method also had issues when reading into a 1-byte array, so that was fixed as well.

Added a set of tests that explain the issue and fail before the fix, but succeeds after.

These are clearly edge case, but if anyone passes a compressed stream to another library, even these edge cases should work correctly.

### CODE REVIEW

File: src/SharpCompress/Providers/Default/GZipCompressionProvider.cs [hunk 1]
```diff
@@ -20,7 +20,13 @@ public sealed class GZipCompressionProvider : CompressionProviderBase
     public override Stream CreateCompressStream(Stream destination, int compressionLevel)
     {
         var level = (CompressionLevel)compressionLevel;
-        return new GZipStream(destination, CompressionMode.Compress, level, Encoding.UTF8);
+        return new GZipStream(
+            destination,
+            CompressionMode.Compress,
+            level,
+            leaveOpen: true,
```
[the remark this answers is not in the document]
Comment (kenkendk) [reply src/SharpCompress/Providers/Default/GZipCompressionProvider.cs]: Fixed

File: src/SharpCompress/IO/IStreamStack.cs [hunk 2]
```diff
@@ -49,14 +49,20 @@ public static Stream GetRootStream(this IStreamStack stack)
...
             {
+                // Check if stream supports seeking before attempting rewind
+                if (!sharpCompressStream.CanSeek)
+                {
+                    return false;
+                }
+
                 // Try to rewind within the buffer. If the position is outside the buffered
                 // region, silently ignore (matching release behavior where streams without
                 // buffering simply didn't rewind).
```
[the remark this answers is not in the document]
Comment (kenkendk) [reply src/SharpCompress/IO/IStreamStack.cs]: Updated comment

File: tests/SharpCompress.Test/Zip/ZipCompressionRoundtripTests.cs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+using SharpCompress.Readers;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+using Xunit;
+
+namespace SharpCompress.Test.Zip;
+
+public class ZipCompressionRoundtripTests : TestBase
+{
+    private const int TestSize = 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (kenkendk) [reply tests/SharpCompress.Test/Zip/ZipCompressionRoundtripTests.cs]: Reduced to single pass and 100kb

**Comment by adamhathcock** [comment]
Thanks for finding this.  I'm mainly nervous around its usage and the using Read() on an asynchronous stream.

I have tests that only use the async version to prove things.

I'm wondering if it's best to just have the ReadByte throw an exception and have the Read and ReadAsync methods work with single bytes

**Comment by kenkendk** [comment]
@adamhathcock Not sure I understand the concern. The `ReadByte()` method calls `Read()` and both are sync methods, so no change there?

**Comment by adamhathcock** [comment]
but there isn't a ReadByte for the async path.  Since ReadByte reuses the sync path, it would fail on asynchronous operations/tests.

**Comment by adamhathcock** [comment]
Added the bug here: https://github.com/adamhathcock/sharpcompress/pull/1346

Tempted to just make ZLibStream.ReadByte return -1
